tensorflow-keras/src/main/java/org/tensorflow/keras/optimizers/AdaDelta.java

```diff
@@ -0,0 +1,200 @@
+   *
+   * @param tf the tensorflow Ops
+   * @param learningRate The learning rate
+   */
+  public AdaDelta(Ops tf, float learningRate) {
+    this(tf, learningRate, RHO_DEFAULT, EPSILON_DEFAULT);
+  }
+
+  /**
+   * Create an Adadelta optimizer with default rho=0.95F, and epsilon=1e-7F
+   *
+   * @param tf the tensorflow Ops
+   * @param name the name of the Optimizer, defaults to "Adadelta"
+   * @param learningRate The learning rate
+   */
+  public AdaDelta(Ops tf, String name, float learningRate) {
+    this(tf, name, learningRate, RHO_DEFAULT, EPSILON_DEFAULT);
+  }
+
+  /**
+   * Create an Adadelta optimizer with default name="Adadelta",
+   *
+   * @param tf the tensorflow Ops
+   * @param learningRate The learning rate
+   * @param rho The decay rate.
+   * @param epsilon A constant epsilon used to better conditioning the grad update.
+   */
+  public AdaDelta(Ops tf, float learningRate, float rho, float epsilon) {
+    super(assertGraph(tf), learningRate, rho, epsilon);
+    initConfig(learningRate, rho, epsilon);
+  }
+
+  /**
+   * Create an Adadelta optimizer
+   *
+   * @param tf the tensorflow Ops
+   * @param name the name of the Optimizer, defaults to "Adadelta"
+   * @param learningRate The learning rate
+   * @param rho The decay rate.
+   * @param epsilon A constant epsilon used to better conditioning the grad update.
+   */
+  public AdaDelta(Ops tf, String name, float learningRate, float rho, float epsilon) {
+    super(assertGraph(tf), name, learningRate, rho, epsilon);
+    initConfig(learningRate, rho, epsilon);
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  protected Optional<Op> prepare(String name) {
+    switch (initializers.size()) {
+      case 0:
+        return Optional.empty();
+      case 1:
+        return Optional.of(initializers.get(0));
+      default:
+        return Optional.of( tf.withSubScope(name).withControlDependencies(initializers).noOp());
+    }
+  }
+
+  /**
+   * Create an Adam Optimizer from a config object
+   *
+   * @param tf the tensorflow Ops
+   * @param config a config object to initialize, he config object has keys for "name",
+   *     "learning_rate", "rho" and "epsilon". If a key is missing the default value is used.
+   */
+  public static AdaDelta fromConfig(Ops tf, Map<String, Object> config) {
+    return create(tf, config);
+  }
+
+  /**
+   * Create an Adadelta optimizer
+   *
+   * @param tf the tensorflow Ops
+   * @param config a config object to initialize, the config
+   *     object has keys for "name", "learning_rate", "rho" and "epsilon". If a key is missing the
+   *     default value is used.
+   */
+  public static AdaDelta create(Ops tf, Map<String, Object> config) {
+    String name = (String) config.get(NAME_KEY);
+    float learningRate = (float) config.getOrDefault(LEARNING_RATE_KEY, LEARNING_RATE_DEFAULT);
+    float rho = (float) config.getOrDefault(RHO_RATE_KEY, RHO_DEFAULT);
+    float epsilon = (float) config.getOrDefault(EPSILON_KEY, EPSILON_DEFAULT);
```

Comment:
Can we create a concrete type for the `config` instead of having a map of untyped values? e.g. could it be something like an `AdaDelta.Config` subclass following a building pattern similar to the ops options classes like [this one](https://github.com/tensorflow/java/blob/86840e7766f1f321267c5bf2f2333eefcd659251/tensorflow-core/tensorflow-core-api/src/gen/java/org/tensorflow/op/image/EncodeJpeg.java#L66)? I probably don't understand the use of this `config` format though.

The same comment applies for all other optimizers

Reply:
This was used in Python for "serialization" of the Optimizers. The Config could be saved, then used to restore the state of the Optimizer. The `key` is always a String, but the `value` can be a `String`, a `Float` or `Integer`, (but maybe also a `Double` or `Boolean`), so `Map<String, Object>` should suffice. Since I did this, I have not yet seen where this dictionary is actually used, so I assume the python code expected some user software or tool to use it. I am not sure how this  plays into  saving/resorting the model, as I haven't looked at that logic yet.

We could pull it out for now, then revisit it when we start tackling saving/restoring the Keras model.

Or I can change it to the `options` pattern, as long as there is a way to read/write it, perhaps as a JSON string.

Reply:
I vote for removing the serialization/deserialization support for now and add it later.

Reply:
OK

Reply:
When we return to this we could make the config objects implement `Map<String,Object>` but specialise the types per optimiser and provide specific getters & setters?

Reply:
I have started to write the "options" pattern with internal classes on the Optimizers. Do we want to do this? Some of the optimizers' ctor's are trivial, but then others may have a more significant number of options with Keras specified defaults.

Reply:
I really think we should postpone this feature for now as there is no urgent need to support serialization/deserialization of the configurations, we can then brainstorm on the right approach to take.

Reply:
I have removed it for now and we can revisit later when we look at serializing the Keras Model